### PR TITLE
KCE-85 Update kpsr-core

### PR DESCRIPTION
**Pull request by joseluisbf-kpsr**
[KCE-85]

~**WPI** not all the code was ported, 2/8 tests fail.~

### TODO

a. ~Rename `core/modules/core/include/klepsydra/core/container.h` as `core_container.h`?~ It confuses with `/usr/local/include/klepsydra/sdk/container.h` -> ok, but might break kpsr-robotics and kpsr-tutorial (comment [below](https://github.com/klepsydra-technologies/kpsr-core/pull/99#pullrequestreview-1634954612))
    No occurrences in kpsr-tutorial nor kpsr-robotics.
    -> done, full Klepsydra review in https://github.com/klepsydra-technologies/kpe-core/pull/145#issuecomment-1736201195
b. ~Remove YAML, if possible~ -> done in the past, nothing to do

### Questions

1. ~Include `benchmarks/`? (Moved from `core/tests/performance/`)~ -> yes
3. ~Pending comments at `benchmarks/CMakeLists.txt`:~ -> solved, see [comment](https://github.com/klepsydra-technologies/kpsr-core/pull/99#issuecomment-1737161446) below
    ```cpp
    # "basic_mddlw_test/src/*.cpp"   @todo: Change to use benchmarks
    # "concurrent_mdlw_test/src/*.cpp" @todo: Change to use benchmarks
    ```
    add?
    ```cpp
    "high_performance/src/*.cpp")
    ```
4. ~Add `ci/`?~ -> yes -> done
5. ~Keep TODOs? (e.g. // TODO use lockfree hash maps here. simplify internal structure similar to safe event emitter)~ -> yes, but remove lockfree refs
6. ~Test `BasicEventEmitterTest.SingleEventEmitterTopic` runs forever~. -> ok
7. ~Replace `core/modules/core/include/klepsydra/core/subscription_stats.h` with kpsr-sdk `modules/sdk/include/klepsydra/sdk/subscription_stats.h`? (Almost identical.)~ -> done
8. ~Double copyright in `high_performance/tests/disruptor4cpp_test/src/dummy_sequence_barrier.h` (and more)~ -> Klepsydra removed. -> ok
9. ~In `high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop.h` (and more)~
    ```cpp
    std::thread batchProcessorThread;
    std::future<void> _batchProcessorThreadFuture;
    ```
    ~at kpe-core `batchProcessorThread` is removed (and `_batchProcessorThreadFuture` is renamed to `batchProcessorThread`). How to start the threads with future and without `std::thread` nor `threadpool.h`?~
    > For the moment keep it as is, so that we don't add new complications.

    (see comment [below](https://github.com/klepsydra-technologies/kpsr-core/pull/99#pullrequestreview-1634954612))
10. ~Update/remove `api-doc/`~ -> moxygen generated and outdated, removed

[KCE-85]: https://klepsydra.atlassian.net/browse/KCE-85?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_middleware_provider.h [hunk 1]
```diff
@@ -173,6 +177,7 @@ class DataMultiplexerMiddlewareProvider
     RingBuffer _ringBuffer;
 
     DataMultiplexerPublisher<TEvent, BufferSize> _publisher;
+    std::map<std::string, std::vector<int>> _subscriberCoreMap;
```
Comment (mharshe) [inline high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_middleware_provider.h:180]: not needed.

File: high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_subscriber.h [hunk 2]
```diff
@@ -86,19 +87,19 @@ class DataMultiplexerSubscriber : public EventEmitterSubscriber<TEvent>
             this->_ringBuffer.add_gating_sequences(sequences_to_add);
             this->batchEventProcessor->run();
         })
-        , _batchProcessorThread()
```
Comment (mharshe) [inline high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_subscriber.h:89]: revert this change

File: high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_subscriber.h [hunk 3]
```diff
@@ -86,19 +87,19 @@ class DataMultiplexerSubscriber : public EventEmitterSubscriber<TEvent>
...
         , _batchProcessorThreadFuture(_batchProcessorTask.get_future())
+        , _cpuAffinity(cpuAffinity)
```
Comment (mharshe) [inline high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_subscriber.h:91]: remove affinity

File: high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop.h [hunk 4]
```diff
@@ -70,27 +70,31 @@ class EventLoop
     EventLoop(std::shared_ptr<EventEmitterInterface<EventloopDataWrapper>> &externalEventEmitter,
               RingBuffer &ringBuffer,
               const std::string &name,
-              long timeoutUS = EVENT_LOOP_START_TIMEOUT_MICROSEC)
+              const long timeoutUS = EVENT_LOOP_START_TIMEOUT_MICROSEC,
+              const std::vector<int> &cpuAffinity = {})
```
Comment (mharshe) [inline high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop.h:74]: remove affinity

File: high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop.h [hunk 5]
```diff
@@ -70,27 +70,31 @@ class EventLoop
...
         : _name(name)
         , _threadName(std::to_string(BufferSize) + "_" + name)
         , _ringBuffer(ringBuffer)
         , _eventHandler(externalEventEmitter)
         , _isStarted(false)
         , _eventLoopTask([this] {
+            spdlog::debug("kpsr::high_perf::EventLoop::_eventLoopTask begin");
```
Comment (mharshe) [inline high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop.h:81]:
```suggestion
            spdlog::trace("kpsr::high_perf::EventLoop::_eventLoopTask begin");
```

File: high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop.h [hunk 6]
```diff
@@ -70,27 +70,31 @@ class EventLoop
...
             std::vector<disruptor4cpp::sequence *> sequences_to_add;
             sequences_to_add.resize(1);
             sequences_to_add[0] = &batchEventProcessor->get_sequence();
             this->_ringBuffer.add_gating_sequences(sequences_to_add);
             spdlog::debug(EVENT_LOOP_START_MESSAGE);
             this->batchEventProcessor->run();
+            spdlog::debug("kpsr::high_perf::EventLoop::_eventLoopTask end.");
```
Comment (mharshe) [inline high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop.h:88]:
```suggestion
            spdlog::trace("kpsr::high_perf::EventLoop::_eventLoopTask end.");
```

File: high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop_middleware_provider.h [hunk 7]
```diff
@@ -79,12 +79,12 @@ class EventLoopMiddlewareProvider
                                 const std::string &name = "kpsr_EL",
                                 const long timeoutUS = EVENT_LOOP_START_TIMEOUT_MICROSEC,
                                 const std::vector<int> &cpuAffinity = {},
-                                EventEmitterType eventEmitterType = EventEmitterType::SAFE)
+                                EventEmitterType eventEmitterType = EventEmitterType::UNSAFE_MULTI)
```
Comment (mharshe) [inline high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop_middleware_provider.h:82]: default should be safe

File: high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop_middleware_provider.h [hunk 8]
```diff
@@ -131,7 +134,7 @@ class EventLoopMiddlewareProvider
      * @return
      */
     Subscriber<T> *getSubscriber(const std::string &eventName,
-                                 EventEmitterType eventEmitterType = EventEmitterType::SAFE)
+                                 EventEmitterType eventEmitterType = EventEmitterType::UNSAFE_MULTI)
```
Comment (mharshe) [inline high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop_middleware_provider.h:137]: default safe

File: high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop_scheduler.h [hunk 9]
```diff
@@ -46,8 +46,12 @@ class EventLoopScheduler : public Scheduler
     void startScheduledTask(const std::string &name,
                             int after,
                             bool repeat,
-                            std::function<void()> function) override;
-    void startScheduledService(int after, bool repeat, Service *service) override;
+                            std::function<void()> task,
+                            const std::vector<int> &cpuAffinity = {}) override;
```
Comment (mharshe) [inline high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop_scheduler.h:50]: remove affinity

File: high_performance/tests/high_performance_tests/src/high_perf_middlware_test.cpp [hunk 10]
```diff
@@ -211,13 +212,13 @@ TEST_P(DataMultiplexerMiddlewareConstructorTest, NominalCaseSharedPtr)
     ASSERT_EQ(DataMultiplexerTestEvent::copyInvokations, 0);
 
     auto publisher = provider.getPublisher();
-    auto subscriber = provider.getSubscriber("testSubscriber");
+    auto subscriber = provider.getSubscriber("testSubscriber", {0});
```
Comment (mharshe) [inline high_performance/tests/high_performance_tests/src/high_perf_middlware_test.cpp:215]: removing affinity will allow you to revert this change

**Review by mharshe — COMMENTED** [review]
I don't think the performance and the benchmark modules should be deleted.

File: .github/workflows/push.yml [hunk 11]
```diff
@@ -38,7 +38,7 @@ jobs:
         run: |
```
Comment (mharshe) [inline .github/workflows/push.yml:38]: CI should have a step before this to clone and install KlepsydraSDK

File: CMakeLists.txt [hunk 12]
```diff
@@ -238,46 +262,43 @@ endif()
...
-    set(KPSR_ZMQ_DEP "find_dependency(cppzmq 4.3 QUIET)")
+    set(KPSR_ZMQ_DEP "find_dependency(cppzmq 4.3 QUIET)
+")
+    string(APPEND KPSR_CHECK_DEPS ${KPSR_ZMQ_DEP})
 endif()
 
-set(KPSR_CHECK_DEPS
-    "include(CMakeFindDependencyMacro)
-${KPSR_ZMQ_DEP}
+if(KPSR_COMPILE_SPDLOG)
```
[the remark this answers is not in the document]
Comment (joseluisbf-kpsr) [reply CMakeLists.txt:284]: Fixed at kpsr-build

File: README.md [hunk 13]
```diff
@@ -58,45 +54,27 @@ This will install the klespydra core in default locations:
...
-* -DKPSR_ENABLE_TESTS=true for building unit tests (default true)
+* `-DCMAKE_INSTALL_PREFIX` for specifying the Klepsydra installation location (`/usr/local` by default)
+* `-DKPSR_WITH_DOXYGEN` to allow generation of documentation
+* `-DGTEST_PATH` for the Google test path (default is `./thirdparties/googletest/`)
+* `-DTHIRDPARTIES_PATH` for the ConcurrentQueue and Cereal path (default is `./thirdparties/`)
+* `-DKPSR_WITH_ZMQ=true` for building the ZeroMQ binding
+* `-DKPSR_TEST_PERFORMANCE=true` for building the performance tests. Needs Google [benchmark](https://github.com/google/benchmark)
+* `-DKPSR_WITH_SOCKET` for building SOCKET binding
+* `-DKPSR_ENABLE_EXAMPLES=true` for building examples files (default false)
+* `-DKPSR_ENABLE_TESTS=true` for building unit tests (default true)
```
[the remark this answers is not in the document]
Comment (joseluisbf-kpsr) [reply README.md:66]: MINIMAL_BUILD doesn't change KPSR_ENABLE_EXAMPLES (default false), but it's cited in the description, ok?

File: core/modules/ros_mdlw/kpsr_ros_core/include/kpsr_ros_core/ros_env.h
[the remark this answers is not in the document]
Comment (joseluisbf-kpsr) [reply core/modules/ros_mdlw/kpsr_ros_core/include/kpsr_ros_core/ros_env.h:1]: TODO in kpsr-core

File: core/modules/zmq_mdlw/include/klepsydra/zmq_core/binary_to_zmq_channel.h [hunk 14]
```diff
@@ -49,24 +50,26 @@ class BinaryToZMQChannel : public ObjectPoolPublisher<Base>
...
+                                    std::move(finalizerFunction))
         , _publisher(publisher)
         , _topic(topic)
     {}
 
 protected:
     void internalPublish(std::shared_ptr<const Base> event) override
     {
-        std::stringbuf *buffer = (std::stringbuf *) (*event);
+        auto *buffer = (std::stringbuf *) (*event);
```
[the remark this answers is not in the document]
Comment (joseluisbf-kpsr) [reply core/modules/zmq_mdlw/include/klepsydra/zmq_core/binary_to_zmq_channel.h:72]: `dynamic_cast` instead, ported to kpe-core too.

File: core/modules/zmq_mdlw/include/klepsydra/zmq_core/binary_to_zmq_channel.h [hunk 15]
```diff
@@ -49,24 +50,26 @@ class BinaryToZMQChannel : public ObjectPoolPublisher<Base>
...
-                       zmq::socket_t &publisher)
+                       zmq::socket_t &publisher,
+                       std::function<void(Base &)> finalizerFunction)
         : ObjectPoolPublisher<Base>(container,
                                     topic,
                                     "ZMQ_BINARY",
                                     std::max(1, poolSize),
-                                    initializerFunction,
-                                    nullptr)
+                                    std::move(initializerFunction),
```
[the remark this answers is not in the document]
Comment (joseluisbf-kpsr) [reply core/modules/zmq_mdlw/include/klepsydra/zmq_core/binary_to_zmq_channel.h:62]: image
Part of the ZMQ tests memory leaks fix.

File: core/tests/core_tests/src/service_stats_test.cpp
[the remark this answers is not in the document]
Comment (joseluisbf-kpsr) [reply core/tests/core_tests/src/service_stats_test.cpp:1]: Removed. `service_stats.cpp` at kpsr-sdk.

File: high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_listener.h [hunk 16]
```diff
@@ -115,7 +119,8 @@ class DataMultiplexerListener
     std::packaged_task<void()> _batchProcessorTask;
     std::thread _batchProcessorThread;
     std::future<void> _batchProcessorThreadFuture;
-    long _timeoutUs;
+    const std::vector<int> _cpuAffinity;
```
[the remark this answers is not in the document]
Comment (joseluisbf-kpsr) [reply high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_listener.h:122]: File reverted.

File: high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_listener.h [hunk 17]
```diff
@@ -56,7 +56,8 @@ class DataMultiplexerListener
     DataMultiplexerListener(const std::function<void(TEvent)> &listener,
                             RingBuffer &ringBuffer,
                             const std::shared_ptr<SubscriptionStats> &listenerStat,
-                            long timeoutMS = MULTIPLEXER_START_TIMEOUT_MICROSEC / 1000)
+                            const long timeoutUS = MULTIPLEXER_START_TIMEOUT_MICROSEC,
+                            const std::vector<int> &cpuAffinity = {})
```
[the remark this answers is not in the document]
Comment (joseluisbf-kpsr) [reply high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_listener.h:60]: File reverted.

File: high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_listener.h
[the remark this answers is not in the document]
Comment (joseluisbf-kpsr) [reply high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_listener.h:1]: File reverted.

File: high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_middleware_provider.h [hunk 18]
```diff
@@ -123,17 +124,23 @@ class DataMultiplexerMiddlewareProvider
      * @brief getSubscriber
      * @return
      */
-    Subscriber<TEvent> *getSubscriber(const std::string &subscriberName)
+    Subscriber<TEvent> *getSubscriber(const std::string &subscriberName,
+                                      const std::vector<int> &cpuAffinity = {})
```
[the remark this answers is not in the document]
Comment (joseluisbf-kpsr) [reply high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_middleware_provider.h:128]: File reverted (but SDK header).

File: high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_middleware_provider.h (hunk 1 above)
Comment (joseluisbf-kpsr) [reply high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_middleware_provider.h:180]: File reverted.

File: high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_subscriber.h (hunk 2 above)
Comment (joseluisbf-kpsr) [reply high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_subscriber.h:89]: File reverted (but var rename).

File: high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_subscriber.h (hunk 3 above)
Comment (joseluisbf-kpsr) [reply high_performance/modules/high_performance/include/klepsydra/high_performance/data_multiplexer_subscriber.h:91]: File reverted (but var rename).

File: core/modules/core/include/klepsydra/core/function_stats.h
Comment (mharshe) [inline core/modules/core/include/klepsydra/core/function_stats.h:1]: remove file since it is in sdk

**Review by mharshe — COMMENTED** [review]
> Keep TODOs? (e.g. // TODO use lockfree hash maps here. simplify internal structure similar to safe event emitter)

yes, but remove refs to lockfree

> Rename core/modules/core/include/klepsydra/core/container.h as core_container.h?

Yes I think you can do that, but it might break kpsr-robotics and kpsr-tutorial (which might need updating after this change any way).

> event_loop.h : 
>
> ```
> std::thread batchProcessorThread;
> std::future<void> _batchProcessorThreadFuture;
> ```

Keep these as before. std::thread t( <function>) starts the thread with the function. `packaged_task  sampleTask` can be used to create a task that can be started later (by doing `std::thread t(std::move(sampleTask))`) . `future` is used to get the result of the function. It is used here but probably not necessary since all our tasks return void. For the moment keep it as is, so that we don't add new complications.

File: high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop.h (hunk 5 above)
Comment (joseluisbf-kpsr) [reply high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop.h:81]: Same for end, ported to kpe-core.
`spdlog::debug(EVENT_LOOP_START_MESSAGE)` kept.

File: high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop_middleware_provider.h (hunk 7 above)
Comment (joseluisbf-kpsr) [reply high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop_middleware_provider.h:82]: 21x replacements.

File: .github/workflows/push.yml (hunk 11 above)
Comment (joseluisbf-kpsr) [reply .github/workflows/push.yml:38]: Not modified in this PR, @tomasz3dk?

Comment (mharshe) [reply .github/workflows/push.yml:38]: If you don't add it, the CI will always fail :)

File: high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop_scheduler.h (hunk 9 above)
Comment (joseluisbf-kpsr) [reply high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop_scheduler.h:50]: Removing from [kpst-sdk](https://github.com/klepsydra-technologies/kpsr-sdk/blob/c5d43384bd9a08e5b0501072feb796135a0b91e0/modules/sdk/include/klepsydra/sdk/scheduler.h) too, @mharshe?

File: core/modules/core/include/klepsydra/core/lockfree_hash_table_uint.h
[the remark this answers is not in the document]
Comment (joseluisbf-kpsr) [reply core/modules/core/include/klepsydra/core/lockfree_hash_table_uint.h:1]: only lock_free_stack.h (was already in kpsr-core)

File: core/modules/mem_mdlw/src/basic_scheduler.cpp [hunk 19]
```diff
@@ -13,48 +13,73 @@
...
+            if (_isRunning) {
+                std::this_thread::sleep_for(std::chrono::microseconds(_after));
+                _task();
+            }
         }
-    };
-
-    _thread = std::thread(scheduledTask);
+    });
+    _thread = scheduledTask.get_future();
```
[the remark this answers is not in the document]
Comment (joseluisbf-kpsr) [reply core/modules/mem_mdlw/src/basic_scheduler.cpp]: `std::function<void()> scheduledTask` updated, move added (was `_thread = std::thread(scheduledTask);`).

File: high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop_scheduler.h (hunk 9 above)
Comment (mharshe) [reply high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop_scheduler.h]: yes remove from kpsr-sdk. Didn't realize this before there.

File: core/modules/core/include/klepsydra/core/function_stats.h
Comment (joseluisbf-kpsr) [reply core/modules/core/include/klepsydra/core/function_stats.h:1]: `subscription_stats.h` removed too.

**Comment by pablogh-2000** [comment]
> > Keep TODOs? (e.g. // TODO use lockfree hash maps here. simplify internal structure similar to safe event emitter)
> 
> yes, but remove refs to lockfree
> 
> > Rename core/modules/core/include/klepsydra/core/container.h as core_container.h?
> 
> Yes I think you can do that, but it might break kpsr-robotics and kpsr-tutorial (which might need updating after this change any way).
> 
> > event_loop.h :
> > ```
> > std::thread batchProcessorThread;
> > std::future<void> _batchProcessorThreadFuture;
> > ```
> 
> Keep these as before. `std::thread t( <function>)` starts the thread with the function. `packaged_task sampleTask` can be used to create a task that can be started later (by doing `std::thread t(std::move(sampleTask))`) . `future` is used to get the result of the function. It is used here but probably not necessary since all our tasks return void. For the moment keep it as is, so that we don't add new complications.

For the idea of core_container.h, it will diverge from kpe-core no? In that case we either do it in both or in none of them.

**Comment by joseluisbf-kpsr** [comment]
> For the idea of core_container.h, it will diverge from kpe-core no? In that case we either do it in both or in none of them.

Right, some changes in this PR are applied to simultaneously to kpsr-core and kpe-core.

File: CMakeLists.txt [hunk 20]
```diff
@@ -16,128 +16,137 @@
...
+include(CheckCXXCompilerFlag)
+include(CheckIncludeFileCXX)
+include(GenerateVersion)
+include(SetCxxStandard)
+include(KlepsydraBuild)
+include(KlepsydraLinkerFlags)
+include(ConfigureLinux)
+
+set(PROJ_MAIN_NAME "klepsydra")
+project("kpsr-core")
```
[the remark this answers is not in the document]
Comment (joseluisbf-kpsr) [reply CMakeLists.txt]: DEB calls only "core", I suggest going back to 'kpsr-core'

[the remark this answers is not in the document]
Comment (mharshe) [reply CMakeLists.txt]: ok

File: .github/workflows/push.yml (hunk 11 above)
Comment (joseluisbf-kpsr) [reply .github/workflows/push.yml:50]: Done, very hand-made, please review.

File: .github/workflows/push.yml [hunk 21]
```diff
@@ -29,16 +29,29 @@ jobs:
...
+          make -j$(nproc)
+          make install
+          rm -rf ${GITHUB_WORKSPACE}/kpsr-sdk/
+
       - name: Build
         run: |
           git config --global --add safe.directory ${GITHUB_WORKSPACE}
           mkdir build && cd build
-          cmake -DKPSR_WITH_DOXYGEN=true -DKPSR_WITH_DDS=false -DKPSR_WITH_ZMQ=false -DKPSR_TEST_PERFORMANCE=false -DKPSR_WITH_SOCKET=true -DKPSR_WITH_CODE_METRICS=true -DCMAKE_PREFIX_PATH=/opt/klepsydra/thirdparties -DCMAKE_BUILD_TYPE=Debug ../
+          cmake -DKPSR_WITH_DOXYGEN=true -DKPSR_WITH_ZMQ=false -DKPSR_TEST_PERFORMANCE=false -DKPSR_WITH_SOCKET=true -DKPSR_COMPILE_SPDLOG=false \
```
Comment (joseluisbf-kpsr) [inline .github/workflows/push.yml:53]: Default `-DKPSR_COMPILE_SPDLOG=true` makes the CI fail, includes not finding "spdlog/spdlog.h", why?
Local build ok for both values.

File: .github/workflows/push.yml [hunk 22]
```diff
@@ -29,16 +29,29 @@ jobs:
...
+      - uses: actions/checkout@v4
         with:
           submodules: recursive
           fetch-depth: 0
 
+      - name: Build SDK
+        run: |
+          git clone https://github.com/klepsydra-technologies/kpsr-sdk.git
+          cd kpsr-sdk/
+          git switch KCE-84-Create-kpsr-sdk
```
Comment (joseluisbf-kpsr) [inline .github/workflows/push.yml:41]: remove once SDK is merged

File: .github/workflows/push.yml [hunk 23]
```diff
@@ -29,16 +29,29 @@ jobs:
...
+          make install
+          rm -rf ${GITHUB_WORKSPACE}/kpsr-sdk/
+
       - name: Build
         run: |
           git config --global --add safe.directory ${GITHUB_WORKSPACE}
           mkdir build && cd build
-          cmake -DKPSR_WITH_DOXYGEN=true -DKPSR_WITH_DDS=false -DKPSR_WITH_ZMQ=false -DKPSR_TEST_PERFORMANCE=false -DKPSR_WITH_SOCKET=true -DKPSR_WITH_CODE_METRICS=true -DCMAKE_PREFIX_PATH=/opt/klepsydra/thirdparties -DCMAKE_BUILD_TYPE=Debug ../
+          cmake -DKPSR_WITH_DOXYGEN=true -DKPSR_WITH_ZMQ=false -DKPSR_TEST_PERFORMANCE=false -DKPSR_WITH_SOCKET=true -DKPSR_COMPILE_SPDLOG=false \
+                -DKPSR_WITH_CODE_METRICS=true -DCMAKE_PREFIX_PATH=/opt/klepsydra/thirdparties -DCMAKE_BUILD_TYPE=Debug ../
```
Comment (mharshe) [inline .github/workflows/push.yml:54]: I don't think you need to give this CMAKE_PREFIX_PATH

File: benchmarks/smart_pool/src/smart_pool_benchmark.cpp [hunk 24]
```diff
@@ -141,36 +141,6 @@ BENCHMARK_DEFINE_F(SmartPoolFixture, VectorMultiThreadOperate)(benchmark::State
     }
 }
 
-static void BM_SmartPoolCreateAndAcquireVector(benchmark::State &state)
```
[the remark this answers is not in the document]
Comment (joseluisbf-kpsr) [reply benchmarks/smart_pool/src/smart_pool_benchmark.cpp:144]: This file is identical to kpe-core's (but license). Add them to kpe-core too?

[the remark this answers is not in the document]
Comment (mharshe) [reply benchmarks/smart_pool/src/smart_pool_benchmark.cpp:144]: maybe in kpe-core for later.

File: core/modules/socket_mdlw/include/klepsydra/socket_core/server_socket.h [hunk 25]
```diff
@@ -60,7 +60,7 @@ class ServerSocket : public BaseSocket
      * @details An accepts waits for a connection and returns a
      * socket object that can be used by the client for communication
      */
-    std::shared_ptr<DataSocket> accept(int timeout_us = 1000000);
+    std::shared_ptr<DataSocket> accept(int timeout = 1);
```
[the remark this answers is not in the document]
Comment (joseluisbf-kpsr) [reply core/modules/socket_mdlw/include/klepsydra/socket_core/server_socket.h:63]: fixed on kpe too

File: benchmarks/event_emitter_test/src/event_emitter_benchmark.cpp [hunk 26]
```diff
@@ -15,15 +15,18 @@
...
 
+#include <klepsydra/core/core_container.h>
 #include <klepsydra/core/event_emitter_middleware_provider.h>
 
 #include "test_result_data.h"
 
 static const std::vector<std::tuple<std::string, kpsr::EventEmitterType>> scenarios =
     {{"Single Listener", kpsr::EventEmitterType::UNSAFE_SINGLE},
-     {"Fixed Listeners", kpsr::EventEmitterType::UNSAFE_MULTI},
+     {"Fixed Listeners", kpsr::EventEmitterType::SAFE},
```
[the remark this answers is not in the document]
Comment (joseluisbf-kpsr) [reply benchmarks/event_emitter_test/src/event_emitter_benchmark.cpp:29]: `event_loop_test.cpp` reverted too.

**Comment by pablogh-2000** [comment]
> [KCE-85](https://klepsydra.atlassian.net/browse/KCE-85?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ)
> 
> ~**WPI** not all the code was ported, 2/8 tests fail.~
> ### TODO
> 
> a. Rename `core/modules/core/include/klepsydra/core/container.h` as `core_container.h`? It confuses with `/usr/local/include/klepsydra/sdk/container.h` -> ok, but might break kpsr-robotics and kpsr-tutorial (comment [below](https://github.com/klepsydra-technologies/kpsr-core/pull/99#pullrequestreview-1634954612)) No occurrences in kpsr-tutorial nor kpsr-robotics. b. ~Remove YAML, if possible~ -> done in the past, nothing to do
> ### Questions
> 
>     1. ~Include `benchmarks/`? (Moved from `core/tests/performance/`)~ -> yes
> 
>     2. Pending comments at `benchmarks/CMakeLists.txt`:
>        ```c++
>        # "basic_mddlw_test/src/*.cpp"   @todo: Change to use benchmarks
>        # "concurrent_mdlw_test/src/*.cpp" @todo: Change to use benchmarks
>        ```
>        
>        
>            
>              
>            
>        
>              
>            
>        
>            
>          
>        add?
>        ```c++
>        "high_performance/src/*.cpp")
>        ```
> 
>     3. Add `ci/`?
> 
>     4. ~Keep TODOs? (e.g. // TODO use lockfree hash maps here. simplify internal structure similar to safe event emitter)~ -> yes, but remove lockfree refs
> 
>     5. ~Test `BasicEventEmitterTest.SingleEventEmitterTopic` runs forever~. -> ok
> 
>     6. Replace `core/modules/core/include/klepsydra/core/subscription_stats.h` with kpsr-sdk `modules/sdk/include/klepsydra/sdk/subscription_stats.h`? (Almost identical.)
> 
>     7. Double copyright in `high_performance/tests/disruptor4cpp_test/src/dummy_sequence_barrier.h` (and more) -> Klepsydra removed.
> 
>     8. ~In `high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop.h` (and more)~
>        ```c++
>        std::thread batchProcessorThread;
>        std::future<void> _batchProcessorThreadFuture;
>        ```
>        
>        
>            
>              
>            
>        
>              
>            
>        
>            
>          
>        ~at kpe-core `batchProcessorThread` is removed (and `_batchProcessorThreadFuture` is renamed to `batchProcessorThread`). How to start the threads with future and without `std::thread` nor `threadpool.h`?~
>        > For the moment keep it as is, so that we don't add new complications.
>        
>        
>        (see comment [below](https://github.com/klepsydra-technologies/kpsr-core/pull/99#pullrequestreview-1634954612))
> 
>     9. Update/remove `api-doc/`

My take:
2. I don't understand the question
3. Yes
6. @mharshe I think what @joseluisbf-kpsr said makes sense. We should have only one stats. Do we have the same duplication in kpe-core as well?
7. Yes
9. I don't know this folder.

**Comment by mharshe** [comment]
> > 2. Pending comments at `benchmarks/CMakeLists.txt`:
> >    ```c++
> >    # "basic_mddlw_test/src/*.cpp"   @todo: Change to use benchmarks
> >    # "concurrent_mdlw_test/src/*.cpp" @todo: Change to use benchmarks
> >    ```
> >      
> >    add?
> >    ```c++
> >    "high_performance/src/*.cpp")
> >    ```

This is already added now as far as I see.  So this question is resolved.

> > 6. Replace `core/modules/core/include/klepsydra/core/subscription_stats.h` with kpsr-sdk `modules/sdk/include/klepsydra/sdk/subscription_stats.h`? (Almost identical.)
files from kpe-sdk should not be repeated. I think by now the subscription_stats.h is removed from this PR. It is not present in kpe-core.
> > 
> > 9. Update/remove `api-doc/`
> > ```
>  9. I don't know this folder.
I think this is generated by moxygen.

File: high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop_middleware_provider.h (hunk 7 above)
Comment (joseluisbf-kpsr) [reply high_performance/modules/high_performance/include/klepsydra/high_performance/event_loop_middleware_provider.h]: Wrong massive replacement, reverted below.

**Comment by joseluisbf-kpsr** [comment]
All TODOs and questions addressed.

File: .github/workflows/prcodecheck.yml [hunk 27]
```diff
@@ -49,14 +69,31 @@ jobs:
     container:
       image: ghcr.io/klepsydra-technologies/kpsrbase:1.0.0
     steps:
-      - uses: actions/checkout@v3
+      - name: Clone kpsr-sdk
+        uses: actions/checkout@v4
+        with:
+          repository: klepsydra-technologies/kpsr-sdk
+          ref: KCE-84-Create-kpsr-sdk
```
Comment (tomasz3dk) [inline .github/workflows/prcodecheck.yml:76]: Same as previous comment

**Review by pablogh-2000 — APPROVED** [review]

**Review by pablogh-2000 — COMMENTED** [review]
Same here, create a JIRA to update, and publish automatically the doxygen doc in HTML

**Comment by joseluisbf-kpsr** [comment]
> Same here, create a JIRA to update, and publish automatically the doxygen doc in HTML

[KCE-87]

**Review by mharshe — APPROVED** [review]
